src/node/interfaces.cpp

```diff
@@ -938,9 +938,85 @@ class BlockTemplateImpl : public BlockTemplate
         return chainman().ProcessNewBlock(block_ptr, /*force_processing=*/true, /*min_pow_checked=*/true, /*new_block=*/nullptr);
     }
 
+    std::unique_ptr<BlockTemplate> waitNext(CAmount fee_threshold, MillisecondsDouble timeout) override
+    {
+        // Delay calculating the current template fees, just in case a new block
+        // comes in before the next tick.
+        CAmount current_fees = -1;
+
+        // Alternate waiting for a new tip and checking if fees have risen.
+        // The latter check is expensive so we only run it once per second.
+        auto now{std::chrono::steady_clock::now()};
+        const auto deadline = now + timeout;
+        const MillisecondsDouble tick{1000};
+
+        // Lower than OR equal is used here, so that a fee_threshold of 0 combined with
+        // a timeout of 0 immedidately returns a new template (rather than nullptr).
+        while (now <= deadline) {
+            bool tip_changed{false};
+            {
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    // Although C++17 allows safe comparison of optional<T> with
+                    // another T, we need to wait for m_tip_block to be set AND
+                    // for the value to be different than the current_tip value.
+                    tip_changed = notifications().m_tip_block && notifications().m_tip_block != m_block_template->block.hashPrevBlock;
```

Comment:
If #31346 lands first, `notifications().m_tip_block` could be an `Assert` and the comment removed, see https://github.com/bitcoin/bitcoin/pull/31346#issuecomment-2493956840